lib/semantics/check-do.cc

```diff
@@ -833,6 +834,50 @@ void DoChecker::Leave(const parser::ConnectSpec &connectSpec) {
   }
 }
 
+using ActualArgumentSet = std::set<evaluate::ActualArgumentRef>;
+
+struct CollectActualArgumentsHelper
+  : public evaluate::SetTraverse<CollectActualArgumentsHelper,
+        ActualArgumentSet> {
+  using Base = SetTraverse<CollectActualArgumentsHelper, ActualArgumentSet>;
+  CollectActualArgumentsHelper() : Base{*this} {}
+  using Base::operator();
+  ActualArgumentSet operator()(const evaluate::ActualArgument &arg) const {
+    return ActualArgumentSet{arg};
+  }
+};
+
+template<typename A> ActualArgumentSet CollectActualArguments(const A &x) {
+  return CollectActualArgumentsHelper{}(x);
+}
+
+template ActualArgumentSet CollectActualArguments(const SomeExpr &);
+
+void DoChecker::Leave(const parser::Expr &parserExpr) {
+  if (parserExpr.typedExpr) {
+    if (const SomeExpr * expr{GetExpr(parserExpr)}) {
+      ActualArgumentSet argSet{CollectActualArguments(*expr)};
+      for (const evaluate::ActualArgumentRef &argRef : argSet) {
+        if (const SomeExpr * argExpr{argRef->UnwrapExpr()}) {
+          if (evaluate::IsVariable(*argExpr)) {
+            if (const Symbol * symbol{evaluate::GetFirstSymbol(*argExpr)}) {
```

Comment:
Can this be written as a helper function/lambda returning a symbol for an argExpr which satisfies some conditions?

Reply:
Thanks, Kiran.  In fact, there's already a function (UnwrapWholeSymbolDataRef) that returns a Symbol for an expression that's a variable.  I'll use that function here.

Reply:
In fact, it looks like I can share a lot of the code between the checking for CallStmt and Expr nodes.